### PR TITLE
introduce the various enlightenments necessary for working with statically instrumented binaries.

### PR DESCRIPTION
**Pull request by 0vercl0k**
This PR introduces the ability for afl-fuzz, afl-showmap and winafl-cmin to work with binaries that have been statically instrumented with syzygy's afl mode, and its associated documentation.

Cheers

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -257,15 +276,152 @@ The target function should do these things during its lifetime:
...
+
+* force-decompose: This switch lets you override the decision that [syzygy](https://github.com/google/syzygy/blob/master/syzygy/pe/pe_transform_policy.cc#L175) makes when evaluating
+if a function is safe to decompose. If you turn on this flag, your instrumentation coverage will be
+higher but you might end-up in an executable that *crashes* in weird ways. Only use if you know what you
+are doing.
+
+* multithread: This switch turns on the thread-safe instrumentation. The major difference with the single
+thread instrumentation is that `__afl_prev_loc` will be stored in a TLS slot.
+
+### How to write a target function
```
Comment (ivanfratric) [inline README.md]: Probably "How to write a target function" and "Building instrument.exe" should be before "Instrumenting a target"

File: README.md [hunk 2]
```diff
@@ -257,15 +276,152 @@ The target function should do these things during its lifetime:
...
 
-8) FAQ
-------
+In order to prepare your target, you need to invoke `__afl_persistent_loop` function like in `test_static.cpp`:
 
+```
```
Comment (ivanfratric) [inline README.md]: add

`#include afl-staticinstr.h`

File: README.md [hunk 3]
```diff
@@ -257,15 +276,152 @@ The target function should do these things during its lifetime:
...
+You can invoke AFL tools with the flag `-Y` to enable the static instrumentation mode during fuzzing, corpus minimizing or during test-case minimizing:
+
+```
+afl-fuzz.exe -Y -i minset -o o1 -t 10000 -- -fuzz_iterations 5000 -- test_static.instr.exe @@
+winafl-cmin.py -Y -t 100000 -i in -o minset -- test_static.instr.exe @@
+```
+
+### Building instrument.exe
+
+In order to clone [syzygy](https://github.com/google/syzygy/)'s repository you can follow the instructions outlined here: [SyzygyDevelopmentGuide](https://github.com/google/syzygy/wiki/SyzygyDevelopmentGuide). Once you have `depot_tools` and the repository cloned, you can compile instrument.exe like this:
```
Comment (ivanfratric) [inline README.md]: I see people already asking about the compiled instrument.exe :-) I should compile and include it in the binaries once I rebuild WinAFL with these changes.

File: README.md [hunk 4]
```diff
@@ -257,15 +276,152 @@ The target function should do these things during its lifetime:
...
 
+```
+int fuzz(int argc, char**argv) {
+  while(__afl_persistent_loop()) {
+    test(argc, argv);
+  }
+  return 1;
+}
+```
+
```
Comment (ivanfratric) [inline README.md]: If there are any compiler/linker options needed for the target to work with Syzygy you should also mention them.

File: README.md [hunk 5]
```diff
@@ -257,15 +276,152 @@ The target function should do these things during its lifetime:
...
+`__afl_persistent_loop`'s implementation lives inside `afl-staticinstr.c` and basically reproduces what the DynamoRIO plugin is doing in `pre_fuzz_handler` and `post_fuzz_handler`. Every points mentioned in "How to select a target function" applies here too.
+
```
Comment (ivanfratric) [inline README.md]: Maybe a separate section for this for visibility, 'Fuzzing a statically instrumented binary' or something similar.

**Comment by ivanfratric** [comment]
Not sure if I'll be able to hop to work to publish the new binary release with this before my Black Hat / DefCon trip and the subsequent vacation. In case I don't make it, I encourage anyone that wishes to use this to compile this revision themselves and report any issues here.

File: README.md (hunk 1 above)
Comment (0vercl0k) [inline README.md]: Done!

File: README.md (hunk 2 above)
Comment (0vercl0k) [inline README.md]: Done!

File: README.md (hunk 3 above)
Comment (0vercl0k) [inline README.md]: That'd be great :).

File: README.md (hunk 4 above)
Comment (0vercl0k) [inline README.md]: Added this at the beginning of "instrumenting a target".

File: README.md (hunk 5 above)
Comment (0vercl0k) [inline README.md]: This is currently in this:
## Statically instrument a binary via [syzygy](https://github.com/google/syzygy)

You can see better the result of the README.md here: https://github.com/0vercl0k/winafl/tree/dev-staticinstrumentation :).

**Comment by 0vercl0k** [comment]
Addressed the first round of feedback Ivan!

Cheers

**Comment by ivanfratric** [comment]
Sorry, had to revert. test_static / afl-staticinstr.c do not build on VS 2010 (I know, pretty old but I've been using it so far and don't have the time to setup a different build environment).

DBG_PRINTEXCEPTION_WIDE_C is undefined and I'm getting a lot of syntax errors.

**Comment by 0vercl0k** [comment]
Sorry for the hassle Ivan - do you have the log somewhere? Also, what
target SDK are you using?

Cheers

2017-07-24 4:54 GMT-07:00 ivanfratric <notifications@github.com>:

**Comment by 0vercl0k** [comment]
I've looked a bit more at this today, and I fixed a couple of things:

- Linking against psapi.lib for older platform (EnumProcessModulesEx / GetModuleBaseName) (https://github.com/0vercl0k/winafl/commit/f2fef1de118a3f127a88b6604523e37f935eeca6#diff-af3b638bc2a3e6c650974192a53c7291R17)

- ifdef'd DBG_PRINTEXCEPTION_WIDE_C as it has been introduced in the Win10 SDK (https://github.com/0vercl0k/winafl/commit/f2fef1de118a3f127a88b6604523e37f935eeca6#diff-6e4dc8332d658b228c2e38c22cecfe47R135)

- Correct the formatting of https://github.com/0vercl0k/winafl/commit/f2fef1de118a3f127a88b6604523e37f935eeca6#diff-6e4dc8332d658b228c2e38c22cecfe47R318 as it made VS2010 unhappy

All the changes are available at https://github.com/0vercl0k/winafl/tree/dev-staticinstrumentation, so let me know if this works better and I'll open another PR :).

Cheers

**Comment by ivanfratric** [comment]
There are still a couple of minor issues that prevent test_static from building on older compiler versions:

 - Declaring variables in loops is not supported in C prior to VC2013 (lots of issues like that in afl-staticinstr.c)
 - test_static.cpp(89): warning C4530: C++ exception handler used, but unwind semantics are not enabled. Specify /EHsc
 - Building the 64-bit version throws afl-staticinstr.h(35): fatal error C1189: #error :  Static instrumentation is only available for 32 bit binaries. Is it possible to modify CMakeLists so that 64-bit test_static won't be built at all rather than throwing an error?

Could you resolve those and submit a new PR?

**Comment by 0vercl0k** [comment]
Thanks for the feedback Ivan - I've addressed the points you raised & the
changes are available here:
https://github.com/0vercl0k/winafl/commit/db3049e0ecc283b8ef65b4591f91b8952d5d2776.
Let me know if this is good and I'll send a new PR!

Cheers

2017-08-10 6:38 GMT-07:00 ivanfratric <notifications@github.com>:

**Comment by ivanfratric** [comment]
Thanks! It builds fine now but I'm having trouble running it on some older configurations. I tried Win10 with VS2015 and Win7 with VS2010 and VS2013. Win10/VS2015 runs fine. However on Win7 with either VS2010 or VS2013 it just locks up when attempting to run the instrumented binary. No idea whether it's the OS or the compiler that's causing the issues. Have you tried compiling/running it on different configurations and any idea what might be causing the issues?

Also on my Windows 7 setup, instrument.exe fails with the following error message:
[0811/161914:ERROR:dia_util.cc(64)] Failed to create Dia object.
I see in https://github.com/google/syzygy/wiki/SyzyASanHowTo that possibly missing msdia120.dll might be an issue, but I do have VS 2013 on that config and it doesn't even work if I copy msdia120.dll in the current directory (also tried with msdia140.dll without success). Any insight?

**Comment by 0vercl0k** [comment]
afl-fuzz locks-up or just the statically instrumented binary? If you set
AFL_NO_SINKHOLE=1, does everything to run properly? I'll investigate this
issue this week-end. I have compiled & tried it on several OSs/compilers
(at least win7, winblue & win10).

For dia, I think you have to register it with 'regsvr32 /s msdia120.dll'
(it's an in-proc COM server IIRC) - does it work better?

Cheers,
Axel

2017-08-11 7:34 GMT-07:00 ivanfratric <notifications@github.com>:

**Comment by ivanfratric** [comment]
I didn't have enough time to figure it out but I think it might have been the compiler (possibly instrument.exe not working correctly with code created by older compilers) because code compiled on my Win10 setup did work on Win7 but the code compiled on my Win7 setup (either VS2010 or VS2013) did not work on either.

I used instrument.exe from https://console.cloud.google.com/storage/browser/syzygy-archive/builds/official/190dbfe74c6f5b5913820fa66d9176877924d7c5/?authuser=0&pli=1 (I planned to include that one in the repo).

I can play with it more next week.

**Comment by 0vercl0k** [comment]
Thanks Ivan!

So I have set-up the following environment: Win7 x86 / VS2010 Express
edition, I have registered the 'msdia140.dll' that ships with the
'instrumenter.exe' binary via 'regsvr32 /s msdia140.dll'. I have been able
to instrument test_static directly from the VM.

I have also tried to launch a small fuzzing job, and indeed it'd fail - I
have investigated and the issue was that the '%zu' flag isn't well
understood by the old compiler/crt, so I've replaced it with a simple '%u'
(static instrumentation implies 32bits, so sizeof(size_t) ==
sizeof(uint32_t) anyway) (
https://github.com/0vercl0k/winafl/blob/dev-staticinstrumentation/afl-staticinstr.c#L318
).

I have also added a small section on how to register msdia140.dll (might
want to include it in the binaries too?):
https://github.com/0vercl0k/winafl/tree/dev-staticinstrumentation#registering-msdia140
.

Hopefully it'll now work perfectly everywhere :).

Cheers

2017-08-12 2:03 GMT-07:00 ivanfratric <notifications@github.com>:

**Comment by ivanfratric** [comment]
Yes, works perfectly for me now, please submit a PR!

**Comment by 0vercl0k** [comment]
Done :).

**Pull request merged**